uqcsbot/haiku.py

```diff
@@ -1,55 +1,103 @@
+from yaml import load, Loader
+import random
+
 import discord
 from discord import app_commands
 from discord.ext import commands
 
 from uqcsbot.bot import UQCSBot
+from uqcsbot.utils.err_log_utils import FatalErrorWithLog
+
+SYLLABLE_RULES_PATH: Final[str] = "uqcsbot/static/syllable_rules.yaml"
+ALLOWED_CHANNEL_NAMES: Final[List[str]] = [
+    "banter",
+    "bot-testing",
+    "dating",
+    "food",
+    "general",
+    "memes",
+    "yelling",
+]
+YELLING_CHANNEL_NAME: Final[str] = "yelling"
+HAIKU_BASE_PROBABILITY: float = 0.5
+# How much "more likely" (as determined by _increase_probability) a haiku is if it has punctuation at the end of a line.
+HAIKU_PUNCTUATION_PROBABILITY_INCREASE: float = 1.6
+# Words that increase the probability of a haiku, and the amount they increase the probability by (as determined by _increase_probability)
+HAIKU_FAVOURITE_WORD_LIST: Dict[str, float] = {
+    "haiku": 6,
+    "haikus": 6,
+    "syllable": 4,
+    "word": 1.6,
+    "words": 1.6,
+    "poem": 2,
+    "poems": 2,
+}
+
+# The following should be treated like constants after they are loaded in
+# Affixes should contain all prefix, suffix and infix lists as tuples, as it is easier to work with beginswith and endswith
+affixes: Dict[str, Tuple[str, ...]] = {}
+# Words that are excluded from the syllable counting process and instead have a custom syllable count (like acronyms)
+syllable_exceptions: Dict[str, int] = {}
+# Accents and "equivalent" characters that they should be replaced with for syllable counting purposes
+accent_replacements: Dict[str, str] = {}
+
+with open(SYLLABLE_RULES_PATH, "r", encoding="utf-8") as syllable_rules_file:
+    syllable_rules = load(syllable_rules_file, Loader=Loader)
+# beginswith and endswith both require tuples, so turn all lists into tuples
+for rule_name, rule_specification in syllable_rules.items():
+    if isinstance(rule_specification, list):
+        affixes[rule_name] = tuple(rule_specification)  # type: ignore
+    elif isinstance(rule_specification, dict):
+        match rule_name:
+            case "exceptions":
+                syllable_exceptions = rule_specification
+            case "accents":
+                accent_replacements = rule_specification
+            case _:
+                # We will catch this on __init__ of the cog. We cannot deal with this error now via FatalErrorWithLog, as the bot may not have loaded enough
+                pass
 
 
 class Haiku(commands.Cog):
     """
     Trys to find Haiku messages in certain channels, and respond "Nice haiku" if it finds one
     """
 
-    ALLOWED_CHANNEL_NAMES = [
-        "banter",
-        "bot-testing",
-        "dating",
-        "food",
-        "general",
-        "memes",
-        "yelling",
-    ]
-    YELLING_CHANNEL_NAME = "yelling"
-
     def __init__(self, bot: UQCSBot):
         self.bot = bot
+        if affixes == {} or syllable_exceptions == {} or accent_replacements == {}:
+            raise FatalErrorWithLog(
```

Comment:
fwiw `FatalErrorWithLog` isn't designed for `__init__` contexts, as it needs the admin-alerts channel (which isn't available until ready).

Also, generally, errors in `__init__` context are fatal, because discord.py hasn't setup its callback failure recovery systems around it. Raising a `FatalErrorWithLog` here is effectively the same as a `RuntimeError` in terms of output.

Reply:
On a related note, there's a bit of a mix of errors going on here - due to the "global" (i.e. outside Haiku class) handling of the file stuff, if the file is straight up missing the the bot will also refuse to start (`FileNotFoundError`), but if the file is there and just incomplete the bot will refuse to start (`FatalErrorWithLog`) from a completely different context & with a different exception.

I'd recommend replacing this with a completely new system that just throws all the "can't figure out what's going on with syllable rules" exceptions from one spot. Discussion to happen on discord about `on_ready` vs `__init__` shortly, because I believe jimmy has opinions.